src/mongo/delegates/Updates.class.php

```diff
@@ -478,61 +479,87 @@ protected function applyChangeSet(\Tripod\ChangeSet $cs, $originalCBDs, $context
-
-                    if(!in_array($predicate, $subjectsAndPredicatesOfChange[$subjectOfChange]))
+                    // remove
+                    if (isset($additionsRemovals["removals"]))
                     {
-                        $subjectsAndPredicatesOfChange[$subjectOfChange][] = $predicate;
+                        $elemsToRemove = array();
+                        foreach ($additionsRemovals["removals"] as $removal)
+                        {
+                            foreach($removal as $valueType => $v) { // should only have one k/v
+                                $found = false;
+                                for ($i=0; $i<count($valueObject);$i++) {
+                                    if (isset($valueObject[$i]) && $valueObject[$i][$valueType] == $v)
+                                    {
+                                        // remove $vo
+                                        $elemsToRemove[] = $i;
+                                        $found = true;
+                                        break;
+                                    }
+                                }
+                                if (!$found) {
+                                    $this->errorLog("Removal value {$subjectOfChange} {$predicate} {$v} does not appear in target document to be updated",array("doc"=>$doc));
+                                    throw new \Exception("Removal value {$subjectOfChange} {$predicate} {$v} does not appear in target document to be updated");
+                                }
+                            }
+                        }
+                        if (count($elemsToRemove)>0)
+                        {
+                            foreach($elemsToRemove as $elem)
+                            {
+                                unset($valueObject[$elem]);
+                            }
+                            $valueObject = array_values($valueObject); // renumbers array after unsets
+                        }
                     }
 
-                    $object = $cs->get_subject_property_values($r["value"],$this->labeller->qname_to_uri("rdf:object"));
-
-                    $isUri = ($object[0]['type']=="uri");
-
-                    if ($isUri)
+                    if (count($valueObject)>0)
                     {
-                        $targetGraph->add_resource_triple($subjectOfChange,$predicate,$object[0]["value"]);
+                        // unique value object
+                        $valueObject = array_map("unserialize", array_unique(array_map("serialize", $valueObject)));
+
+                        if (count($valueObject)==1)
+                        {
+                            $valueObject = $valueObject[0]; // un-array if only one value
+                        }
+                        $this->addOperatorToChange($mongoUpdateOperations,'$set',array($nsPredicate=>$valueObject));
```

Comment:
I kind of feel we should make these operations available as constants.  ` "$set" ` and ` '$set' ` seem error prone to me and we can pretty easily avoid this with something like ` OPERATION_SET `.